### PR TITLE
first commit for project structure

### PR DESCRIPTION
**Pull request by andreic92**
https://jira.ft.com/browse/FDC-91

### CODE REVIEW

File: _ft/api.yml [hunk 1]
```diff
@@ -0,0 +1,138 @@
+swagger: "2.0"
+
+info:
+  title: Varnish Purger
+  description: >
+    This service is used for purging the varnish cache during information like things and poeple are stored
```
Comment (MihaiSorinMoisa) [inline _ft/api.yml]: There is a typo in `people`

File: .circleci/config.yml [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+          command: docker build .
+  dredd:
+    working_directory: /go/src/github.com/Financial-Times/varnish-purger
+    docker:
+      - image: bankrs/golang-dredd:go1.10.0-dredd5.0.0
+        environment:
+          GOPATH: /go
+          K8S_CONFIG_FILE: ./_ft/test_k8s_file
+          LOGGING_LEVEL: warn
+      - image: peteclarkft/ersatz:stable
```
Comment (tamas-molnar) [inline .circleci/config.yml]: Could we have this image in the official FT DH account + versioned please?

File: README.md [hunk 3]
```diff
@@ -1,2 +1,76 @@
...
+
+## Introduction
+
+This service is used for purging the varnish cache during the storing information like `things` and `people`. This is needed in order to ensure that up-to-date data is provided when the endpoints like `/things` or `/people` are accessed.
+
+## Installation
+      
+Download the source code, dependencies and test dependencies:
+
+        go get -u github.com/kardianos/govendor
```
Comment (tamas-molnar) [inline README.md]: It's dep now.

File: README.md [hunk 4]
```diff
@@ -1,2 +1,76 @@
...
+
+2. Run the binary (using the `help` flag to see the available optional arguments):
+
+        $GOPATH/bin/varnish-purger [--help]
+
+Options:
+
+        --app-system-code="varnish-purger"            System Code of the application ($APP_SYSTEM_CODE)
+        --app-name="varnish-purger"                   Application name ($APP_NAME)
+        --port="8080"                                           Port to listen on ($APP_PORT)
```
Comment (tamas-molnar) [inline README.md]: port line spacing is a bit off

File: README.md [hunk 5]
```diff
@@ -1,2 +1,76 @@
...
+## Healthchecks
+Admin endpoints are:
+
+`/__gtg` - returns 503 if any of the checks executed at the /__health endpoint returns false
+
+`/__health`
+
+Checks if:
+* varnish instances are discovered
+* (maybe a healthcheck onto varnish pods??? if it's possible)
```
Comment (tamas-molnar) [inline README.md]: Can this line be removed?

File: main.go [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+		EnvVar: "LOGGING_LEVEL",
+	})
+
+	log.InitLogger(*appName, *logLevel)
+	log.Infof("[Startup] varnish-purger is starting ")
+
+	app.Action = func() {
+		log.Infof("System code: %s, App Name: %s, Port: %s", *appSystemCode, *appName, *port)
+
+		// create app parameters object
```
Comment (tamas-molnar) [inline main.go]: Do we really need these comments here? Code is descriptive enough, which is a good thing.

File: .circleci/config.yml (hunk 2 above)
Comment (andreic92) [reply .circleci/config.yml]: Maybe @peteclark-ft can help us with this one.

File: .circleci/config.yml [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+    steps:
+      - checkout
+      - run:
+          name: Load ersatz-fixtures.yml to ersatz image
+          command: "curl -X POST --data-binary @_ft/ersatz-fixtures.yml -H \"Content-type: text/x-yaml\" http://localhost:9000/__configure"
```
Comment (tosan88) [inline .circleci/config.yml]: Nice!

File: handler.go [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+	targets, ok := queryParams["target"]
+	if !ok {
+		writer.WriteHeader(http.StatusBadRequest)
+		return
+	}
+
+	for _, target := range targets {
+		err := handler.purgeRequest(target)
+		if err != nil {
+			log.WithError(err).WithField("target", target).Error("Error during purging")
```
Comment (tosan88) [inline handler.go]: Could you please maybe log the transaction id as well?

File: handler.go [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+			Timeout: time.Second * 5,
+		}
+		req, err := http.NewRequest("PURGE", finalURL, nil)
+		if err != nil {
+			return err
+		}
+		resp, err := client.Do(req)
+		if err != nil {
+			return err
+		}
```
Comment (tosan88) [inline handler.go]: As a safety net could you please close the response body and eventually discard it's content? I'm thinking something like this: https://github.com/Financial-Times/composite-orgs-transformer/blob/9d2577256aa16b197d0237b90106ad1a9c0f47be/repository.go#L25

File: handler.go [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+		log.Infof("Response code for %s is %d", finalURL, resp.StatusCode)
```
Comment (tosan88) [inline handler.go]: Could you please maybe log the transaction id as well?

File: main.go [hunk 11]
```diff
@@ -0,0 +1,153 @@
...
+		}()
+
+		waitForSignal()
+		discoverer.Stop()
+	}
+
+	return app
+}
+
+func serveEndpoints(appSystemCode string, appName string, port string, requestHandler requestHandler, healthService *HealthService) {
```
Comment (tosan88) [inline main.go]: I think some parameters are not needed

File: main_test.go [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+
+func TestMainApp(t *testing.T) {
+	app := initApp()
+
+	go func() {
+		app.Run([]string{"--log-level=warn", "--k8s-config-file=_ft/test_k8s_file"})
+	}()
+
+	<-time.After(time.Second * 5)
+	syscall.Kill(syscall.Getpid(), syscall.SIGINT)
```
Comment (tosan88) [inline main_test.go]: Oh, windows is not supporting this call :-/

File: README.md [hunk 13]
```diff
@@ -1,2 +1,84 @@
...
+
+        $GOPATH/bin/varnish-purger [--help]
+
+Options:
+
+        --app-system-code         System Code of the application (env $APP_SYSTEM_CODE) (default "varnish-purger")
+        --app-name                Application name (env $APP_NAME) (default "varnish-purger")
+        --port                    Port to listen on (env $APP_PORT) (default "8080")
+        --target-label-selector   Label selector used for selecting pods (env $TARGET_LABEL_SELECTOR) (default "app in (delivery-varnish)")
+        --k8s-config-file         Path to the k8s config file used for connecting to cluster when the service is executed standalone (env $K8S_CONFIG_FILE) (default "/Users/andrei.ciobanu/.kube/kubeconfig-k8s-eu-delivery")
```
Comment (BerniVarga) [inline README.md]: Default value could be changes here - it shouldn't include personal usernames, I think.

File: discover/discoverer.go [hunk 14]
```diff
@@ -0,0 +1,160 @@
...
+	}
+}
+
+// watchForResources is responsible for watching the pod resources.
+// It returnes two values:
+//  - a boolean that represents if the "Stop" method was called
+//  - an error containing the error that caused the return
+func (d *discoverService) watchForResources(ctx context.Context) (closeRequested bool, resultedError error) {
+	closeRequested = false
+	defer func() {
```
Comment (tosan88) [inline discover/discoverer.go]: Is this panic recovery necessary?

File: handler.go [hunk 15]
```diff
@@ -0,0 +1,83 @@
...
+		err := handler.purgeRequest(target, tid)
+		if err != nil {
+			log.
+				WithTransactionID(tid).
+				WithError(err).
+				WithField("target", target).Error("Error during purging")
+			writer.WriteHeader(http.StatusInternalServerError)
+			return
+		}
+		log.WithTransactionID(tid).Infof("Purging for target %s done", target)
```
Comment (tosan88) [inline handler.go]: Varnish also logs the purge so I think logging successful purges are not necessary

File: discover/discoverer.go (hunk 14 above)
Comment (andreic92) [reply discover/discoverer.go]: Theoretically, no, but I just want to be sure that the flow is not broken by an unexpected panic call.
Any suggestion is appreciated 👍 
Maybe "playing" Go defensively is not go idiomatic?

File: handler.go (hunk 9 above)
Comment (tosan88) [reply handler.go]: I see it's done, nice :+1:

File: handler.go (hunk 15 above)
Comment (andreic92) [reply handler.go]: done 👍

File: discover/discoverer.go (hunk 14 above)
Comment (tosan88) [reply discover/discoverer.go]: I just don't see any calls which might produce panic. On the other hand, in a downstream call there's a recovery as well, which makes more sense to me. But I'm neither against it, so I'll leave it to you if you want to keep it or not :)

File: main.go [hunk 16]
```diff
@@ -0,0 +1,153 @@
...
+	targetLabelSelector := app.String(cli.StringOpt{
+		Name:   "target-label-selector",
+		Value:  "app in (delivery-varnish)",
+		Desc:   "Label selector used for selecting pods",
+		EnvVar: "TARGET_LABEL_SELECTOR",
+	})
+
+	k8sConfigFile := app.String(cli.StringOpt{
+		Name:   "k8s-config-file",
+		Value:  "",
```
Comment (BerniVarga) [inline main.go]: Empty line/value can be removed.

File: healthcheck_test.go [hunk 17]
```diff
@@ -0,0 +1,87 @@
...
+		appDescriptionTest,
+		discoverer.K8SCommunicationCheck())
+
+	// Test Health endpoint
+	healthResponseRecorder := httptest.NewRecorder()
+	healthHandler := http.HandlerFunc(fthealth.Handler(healthService))
+	healthHandler.ServeHTTP(healthResponseRecorder, healthRequest)
+	healthActual := healthResponseRecorder.Result()
+
+	// healtcheck assertion
```
Comment (BerniVarga) [inline healthcheck_test.go]: typo

**Pull request merged**